accepted/single-file/design.md

```diff
@@ -66,225 +32,254 @@ Publishing to a single file can be triggered by adding the following property to
+The Windows mapping routines have the following limitations:
+* [CreateFileMapping](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-createfilemappinga) has no option to create a mapping for a part of the file (no offset argument).
+    This means that:
+    * We cannot use the (SEC_IMAGE) attribute to perform automatic section-wise loading (circumventing alignment requirements) of bundled assemblies directly.
+    * Instead we need to map each section independently.
+* [MapViewOfFile]( https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-mapviewoffileex) can only map parts of a file aligned at [memory allocation granularity]( https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/ns-sysinfoapi-system_info), which is 64KB.
+    This means that:
+    * Each section within the assemblies should be aligned at 64KB – which is not guaranteed by the crossgen compiler.
+    * In order to memory-map one embedded assembly at a time, the assemblies in the bundle must be aligned at 64KB boundaries.
+        * A prototype bundler with this behavior for testing purposes is available in this branch: https://github.com/swaroop-sridhar/core-setup/tree/single-exe
+
+We therefore map the assembly and perform an in-memory copy of the sections to appropriate offsets. In the long term, the solution to the mapping problem would involve considerations such as:
+* Compile all assemblies in a version bubble into one PE assembly, with a few aligned sections.
+* Embed the big composite assembly into the host with proper alignment, so that the single-exe bundle can be loaded without copies at run time.
+
+### API Semantics
 
-An app may choose to only embed some files (ex: due to licensing restrictions) and expect to pickup other dependencies from application-launch directory, nuget packages, etc. In order to resolve assemblies and native libraries, the embedded resources are probed first, followed by other probing paths. 
+#### `Assembly.Location`
 
-### New API
+There are a few options to consider for the `Assembly.Location` property of a bundled assembly:
 
-In this section, we propose adding a few APIs to facilitate common operations on bundled-apps.
+* A fixed literal (ex: `null`) indicating that no actual location is available.
+* The simple name of the assembly (with no path).
+* The path of the assembly as if it were not to be packaged into the single-file.
+* A special UNC notation such as `<bundle-path>/:/asm.dll` to denote files that come from the bundle. 
+* A configurable selection of the above, etc.
 
-The binaries that are published in the project are expected to be handled transparently by the host. However, explicit access to the embedded files is useful in situations such as:
+Proposed solution is: `:/<bundle-sub-path>/asm.dll`, where the `:` denotes the location within the bundle. The `:` is not a valid character in file-names on Windows. The `:` is already used as a path-separator on Unix systems.
```

Comment:
I am not sure whether this is a good idea. We are inventing a new custom path format here. People will pass this path to `File.Exist` or `File.Open` half of the time, and the app is still going to be broken. I do not think we will want to quirk all file I/O APIs to deal with this. 

The most appropriate value value to return from `Assembly.Location` is null.

Reply:
What about `Assembly.CodeBase`?

Reply:
`null` is a valid/documented return value of `Assembly.Location`, but pointing at the documentation and telling people that the NuGet package they're using has a bug because it doesn't check for `null` never helped much. Most apps don't check for the null, and returning null means that most apps will hit a `NullReferenceException` somewhere. People don't know how to deal with that when that exception happens in a NuGet package they don't have source code for. We have plenty of issues like that in the dotnet/corert repo (which already has this problem because it chose `null`).

Returning a custom path has the same problem - `Assembly.Location` is typically passed to the Path APIs because people either want the file part or the directory part. Most uses I've seen want the directory part because e.g. they're trying to scan for plugins or find a config file. `Directory.GetFiles(Path.Combine(Path.GetDirectoryName(asm.Location), "*"))` results in an exception that is similarly unhelpful to a `NullReferenceException`.

I'm wondering whether the best course of action would be to just make this throw with a message that has an fwlink to useful guidance. There can be AppContext switches documented at the fwlink that override the throw with a quirk to make the NuGet package work (return a fake location that combines AppContext.BaseDirectory with the module name, or return null).

`Assembly.CodeBase` is getting obsoleted in .NET 5 (see dotnet/runtime#31127). The appropriate behavior would be to throw.

Reply:
We have trained people to handle Assembly.Location returning null gracefully for last 5 years, and number of NuGet packages made the required fixes already. I do not think getting everybody to change again is a good idea.

There are number of other APIs that are problematic for this mode. We plan to deal with these using analyzer that will warn about the use of these problematic APIs.

Reply:
> We plan to deal with these using analyzer that will warn about the use of these problematic APIs.

@jkotas, mind adding details to https://github.com/dotnet/runtime/issues/30740?  @terrajobst is currently using that to seed our plans for what we're actually going to ship in the release.

Reply:
@sbomer @MichalStrehovsky and @vitek-karas are working on this. The current thinking is to have it in the same ILLinker-bases analyzer that detects problematic reflection APIs, etc. #30740 is about Roslyn-based analyzers.

Reply:
> The current thinking is to have it in the same ILLinker-bases analyzer that detects problematic reflection APIs, etc. [dotnet/runtime#30740](https://github.com/dotnet/runtime/issues/30740) is about Roslyn-based analyzers.

Are you saying that the detection will be part of the linker? While that might make sense for code sharing, it's not necessarily a great developer experience b/c Roslyn-based analyzers are running while editing and thus are providing rapid feedback. In cases where there are alternatives we might want to provide a code fixer, which will force some of the detection logic to be part of the IDE experience anyways.

Reply:
I was debating between null and UNC options: 

Establishing a path-like location for the the bundle contents is useful if the app explicitly wants to use `Assembly.LoadFrom()` on one of the bundled assemblies. Otherwise, `Assembly.LoadFrom()` will always be understood to load from the disk, which is one way to define the semantics.

The path notation is also useful to address content (ex: data files) bundled in the app directly. However, if we only support accessing content through extraction, or as PE resources, this use-case is not a motivation for UNC-like paths.

I'll change the semantics to return `null`.